[ISSUE] Issue with `databricks_catalog_workspace_binding Resource` resource

### Configuration
```
resource "databricks_catalog" "manoj-isolated" {
  name           = "manoj-isolated"
  isolation_mode = "ISOLATED"
}

resource "databricks_catalog_workspace_binding" "manoj-isolated" {
  catalog_name = databricks_catalog.manoj-isolated.name
  workspace_id = 123
```

### Expected Behavior
With the above Resource, it should update the already existing catalog to ISOLATED mode 

### Actual Behavior
but  here it just states that 
[ERROR] vertex "databricks_catalog.manoj-isolated" error: cannot create catalog: Catalog 'manoj-isolated' already exists

### Steps to Reproduce
Run the below resource on already existed catalog which is not isolated already, 
databricks_catalog_workspace_binding Resource
The terraform plan will say resource catalog will be created and apply will fail with the above error. Although [API ](https://docs.databricks.com/data-governance/unity-catalog/create-catalogs.html#configuration)
 is working fine to patch the catalog in isolated mdoe.

### Terraform and provider versions
1.20.0

### Debug Output
2023-07-03T14:45:31.129+0530 [DEBUG] ReferenceTransformer: "databricks_catalog.manoj-isolated" references: []
2023-07-03T14:45:31.129+0530 [DEBUG] ReferenceTransformer: "databricks_catalog_workspace_binding.manoj-isolated" references: [databricks_catalog.manoj-isolated (expand) databricks_catalog.manoj-isolated databricks_catalog.manoj-isolated]

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # databricks_catalog.manoj-isolated will be created
  + resource "databricks_catalog" "manoj-isolated" {
      + force_destroy  = false
      + id             = (known after apply)
      + isolation_mode = "ISOLATED"
      + metastore_id   = (known after apply)
      + name           = "manoj-isolated"
      + owner          = (known after apply)
    }

  # databricks_catalog_workspace_binding.manoj-isolated will be created
  + resource "databricks_catalog_workspace_binding" "manoj-isolated" {
      + catalog_name = "manoj-isolated"
      + id           = (known after apply)
      + workspace_id = "123"
    }

Plan: 2 to add, 0 to change, 0 to destroy.
2023-07-03T14:45:31.130+0530 [DEBUG] command: asking for input: "\nDo you want to perform these actions?"

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: 
2023-07-03T14:45:35.222+0530 [INFO]  backend/local: apply calling Apply
2023-07-03T14:45:35.222+0530 [DEBUG] Building and walking apply graph for NormalMode plan
2023-07-03T14:45:35.223+0530 [DEBUG] Resource state not found for node "databricks_catalog.manoj-isolated", instance databricks_catalog.manoj-isolated
2023-07-03T14:45:35.223+0530 [DEBUG] Resource state not found for node "databricks_catalog_workspace_binding.manoj-isolated", instance databricks_catalog_workspace_binding.manoj-isolated
2023-07-03T14:45:35.223+0530 [DEBUG] ProviderTransformer: "databricks_catalog.manoj-isolated (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/databricks/databricks"]
2023-07-03T14:45:35.223+0530 [DEBUG] ProviderTransformer: "databricks_catalog_workspace_binding.manoj-isolated (expand)" (*terraform.nodeExpandApplyableResource) needs provider["registry.terraform.io/databricks/databricks"]
2023-07-03T14:45:35.223+0530 [DEBUG] ProviderTransformer: "databricks_catalog.manoj-isolated" (*terraform.NodeApplyableResourceInstance) needs provider["registry.terraform.io/databricks/databricks"]
2023-07-03T14:45:35.223+0530 [DEBUG] ProviderTransformer: "databricks_catalog_workspace_binding.manoj-isolated" (*terraform.NodeApplyableResourceInstance) needs provider["registry.terraform.io/databricks/databricks"]
2023-07-03T14:45:35.223+0530 [DEBUG] ReferenceTransformer: "databricks_catalog.manoj-isolated (expand)" references: []
2023-07-03T14:45:35.223+0530 [DEBUG] ReferenceTransformer: "databricks_catalog_workspace_binding.manoj-isolated (expand)" references: []
2023-07-03T14:45:35.223+0530 [DEBUG] ReferenceTransformer: "databricks_catalog.manoj-isolated" references: []
2023-07-03T14:45:35.223+0530 [DEBUG] ReferenceTransformer: "databricks_catalog_workspace_binding.manoj-isolated" references: [databricks_catalog.manoj-isolated (expand) databricks_catalog.manoj-isolated databricks_catalog.manoj-isolated]
2023-07-03T14:45:35.223+0530 [DEBUG] ReferenceTransformer: "provider[\"registry.terraform.io/databricks/databricks\"]" references: []
2023-07-03T14:45:35.224+0530 [DEBUG] Starting graph walk: walkApply
2023-07-03T14:45:35.225+0530 [DEBUG] created provider logger: level=debug
2023-07-03T14:45:35.225+0530 [INFO]  provider: configuring client automatic mTLS
2023-07-03T14:45:35.229+0530 [DEBUG] provider: starting plugin: path=.terraform/providers/registry.terraform.io/databricks/databricks/1.20.0/darwin_arm64/terraform-provider-databricks_v1.20.0 args=[.terraform/providers/registry.terraform.io/databricks/databricks/1.20.0/darwin_arm64/terraform-provider-databricks_v1.20.0]
2023-07-03T14:45:35.232+0530 [DEBUG] provider: plugin started: path=.terraform/providers/registry.terraform.io/databricks/databricks/1.20.0/darwin_arm64/terraform-provider-databricks_v1.20.0 pid=10971
2023-07-03T14:45:35.232+0530 [DEBUG] provider: waiting for RPC address: path=.terraform/providers/registry.terraform.io/databricks/databricks/1.20.0/darwin_arm64/terraform-provider-databricks_v1.20.0
2023-07-03T14:45:35.252+0530 [DEBUG] provider.terraform-provider-databricks_v1.20.0: Databricks Terraform Provider
2023-07-03T14:45:35.252+0530 [DEBUG] provider.terraform-provider-databricks_v1.20.0: 
2023-07-03T14:45:35.252+0530 [DEBUG] provider.terraform-provider-databricks_v1.20.0: Version 1.20.0
2023-07-03T14:45:35.252+0530 [DEBUG] provider.terraform-provider-databricks_v1.20.0: 
2023-07-03T14:45:35.252+0530 [DEBUG] provider.terraform-provider-databricks_v1.20.0: https://registry.terraform.io/providers/databricks/databricks/latest/docs
2023-07-03T14:45:35.252+0530 [DEBUG] provider.terraform-provider-databricks_v1.20.0: 
2023-07-03T14:45:35.255+0530 [INFO]  provider.terraform-provider-databricks_v1.20.0: configuring server automatic mTLS: timestamp=2023-07-03T14:45:35.254+0530
2023-07-03T14:45:35.269+0530 [DEBUG] provider: using plugin: version=5
2023-07-03T14:45:35.269+0530 [DEBUG] provider.terraform-provider-databricks_v1.20.0: plugin address: address=/var/folders/6p/5qqx8cqx3d773rw6491tgc100000gp/T/plugin747668516 network=unix timestamp=2023-07-03T14:45:35.269+0530
2023-07-03T14:45:35.290+0530 [WARN]  ValidateProviderConfig from "provider[\"registry.terraform.io/databricks/databricks\"]" changed the config value, but that value is unused
2023-07-03T14:45:35.291+0530 [INFO]  provider.terraform-provider-databricks_v1.20.0: Explicit and implicit attributes: host, token: timestamp=2023-07-03T14:45:35.290+0530
2023-07-03T14:45:35.294+0530 [WARN]  Provider "registry.terraform.io/databricks/databricks" produced an invalid plan for databricks_catalog.manoj-isolated, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .force_destroy: planned value cty.False for a non-computed attribute
databricks_catalog.manoj-isolated: Creating...
2023-07-03T14:45:35.294+0530 [INFO]  Starting apply for databricks_catalog.manoj-isolated
2023-07-03T14:45:35.294+0530 [DEBUG] databricks_catalog.manoj-isolated: applying the planned Create change
2023-07-03T14:45:36.510+0530 [ERROR] provider.terraform-provider-databricks_v1.20.0: Response contains error diagnostic: tf_req_id=9a018884-a782-ea8c-9f30-a56c5426dfbd @module=sdk.proto diagnostic_detail= diagnostic_severity=ERROR diagnostic_summary="cannot create catalog: Catalog 'manoj-isolated' already exists" @caller=/home/runner/work/terraform-provider-databricks/terraform-provider-databricks/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/diag/diagnostics.go:55 tf_resource_type=databricks_catalog tf_rpc=ApplyResourceChange tf_proto_version=5.3 tf_provider_addr=registry.terraform.io/databricks/databricks timestamp=2023-07-03T14:45:36.509+0530
2023-07-03T14:45:36.517+0530 [ERROR] vertex "databricks_catalog.manoj-isolated" error: cannot create catalog: Catalog 'manoj-isolated' already exists

Error: cannot create catalog: Catalog 'manoj-isolated' already exists

  with databricks_catalog.manoj-isolated,
  on bind.tf line 14, in resource "databricks_catalog" "manoj-isolated":
  14: resource "databricks_catalog" "manoj-isolated" {

2023-07-03T14:45:36.523+0530 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2023-07-03T14:45:36.524+0530 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.terraform.io/databricks/databricks/1.20.0/darwin_arm64/terraform-provider-databricks_v1.20.0 pid=10971
2023-07-03T14:45:36.524+0530 [DEBUG] provider: plugin exited

--- a/catalog/resource_catalog.go
+++ b/catalog/resource_catalog.go
@@ -131,7 +131,20 @@ func ResourceCatalog() *schema.Resource {
 			// So if we don't update the field then the requests would be made to old Name which doesn't exists.
 			d.SetId(ci.Name)
 
-			return nil
+			if d.Get("isolation_mode") != "ISOLATED" {
+				return nil
+			}
+			// Bind the current workspace if the catalog is isolated, otherwise the read will fail
+			currentMetastoreAssignment, err := w.Metastores.Current(ctx)
+			if err != nil {
+				return err
+			}
+			_, err = w.WorkspaceBindings.Update(ctx, catalog.UpdateWorkspaceBindings{
+				Name:             ci.Name,
+				AssignWorkspaces: []int64{currentMetastoreAssignment.WorkspaceId},
+			})
+
+			return err
 		},
 		Delete: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
 			w, err := c.WorkspaceClient()

--- a/catalog/resource_catalog.go
+++ b/catalog/resource_catalog.go
@@ -93,15 +93,17 @@ func ResourceCatalog() *schema.Resource {
 			if err != nil {
 				return err
 			}
-			_, err = w.WorkspaceBindings.Update(ctx, catalog.UpdateWorkspaceBindings{
-				Name:             ci.Name,
-				AssignWorkspaces: []int64{currentMetastoreAssignment.WorkspaceId},
+			_, err = w.WorkspaceBindings.UpdateBindings(ctx, catalog.UpdateWorkspaceBindingsParameters{
+				SecurableName: ci.Name,
+				SecurableType: "catalog",
+				Add: []catalog.WorkspaceBinding{
+					{
+						BindingType: catalog.WorkspaceBindingBindingTypeBindingTypeReadWrite,
+						WorkspaceId: currentMetastoreAssignment.WorkspaceId,
+					},
+				},
 			})
-			if err != nil {
-				return err
-			}
-
-			return nil
+			return err
 		},
 		Read: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
 			w, err := c.WorkspaceClient()
@@ -139,11 +141,16 @@ func ResourceCatalog() *schema.Resource {
 			if err != nil {
 				return err
 			}
-			_, err = w.WorkspaceBindings.Update(ctx, catalog.UpdateWorkspaceBindings{
-				Name:             ci.Name,
-				AssignWorkspaces: []int64{currentMetastoreAssignment.WorkspaceId},
+			_, err = w.WorkspaceBindings.UpdateBindings(ctx, catalog.UpdateWorkspaceBindingsParameters{
+				SecurableName: ci.Name,
+				SecurableType: "catalog",
+				Add: []catalog.WorkspaceBinding{
+					{
+						BindingType: catalog.WorkspaceBindingBindingTypeBindingTypeReadWrite,
+						WorkspaceId: currentMetastoreAssignment.WorkspaceId,
+					},
+				},
 			})
-
 			return err
 		},
 		Delete: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {

--- a/catalog/resource_catalog_test.go
+++ b/catalog/resource_catalog_test.go
@@ -518,14 +518,24 @@ func TestCatalogCreateIsolated(t *testing.T) {
 			},
 			{
 				Method:   "PATCH",
-				Resource: "/api/2.1/unity-catalog/workspace-bindings/catalogs/a",
-				ExpectedRequest: catalog.UpdateWorkspaceBindings{
-					Name:             "a",
-					AssignWorkspaces: []int64{123456789101112},
+				Resource: "/api/2.1/unity-catalog/bindings/catalog/a",
+				ExpectedRequest: catalog.UpdateWorkspaceBindingsParameters{
+					SecurableName: "a",
+					SecurableType: "catalog",
+					Add: []catalog.WorkspaceBinding{
+						{
+							WorkspaceId: int64(123456789101112),
+							BindingType: catalog.WorkspaceBindingBindingTypeBindingTypeReadWrite,
+						},
+					},
 				},
-
-				Response: catalog.CurrentWorkspaceBindings{
-					Workspaces: []int64{123456789101112},
+				Response: catalog.WorkspaceBindingsResponse{
+					Bindings: []catalog.WorkspaceBinding{
+						{
+							WorkspaceId: int64(123456789101112),
+							BindingType: catalog.WorkspaceBindingBindingTypeBindingTypeReadWrite,
+						},
+					},
 				},
 			},
 			{

--- a/internal/acceptance/catalog_test.go
+++ b/internal/acceptance/catalog_test.go
@@ -1,0 +1,51 @@
+package acceptance
+
+import (
+	"testing"
+)
+
+func TestUcAccCatalog(t *testing.T) {
+	unityWorkspaceLevel(t, step{
+		Template: `
+		resource "databricks_catalog" "sandbox" {
+			name         = "sandbox{var.RANDOM}"
+			comment      = "this catalog is managed by terraform"
+			properties = {
+				purpose = "testing"
+			}
+		}`,
+	})
+}
+
+func TestUcAccCatalogIsolated(t *testing.T) {
+	unityWorkspaceLevel(t, step{
+		Template: `
+		resource "databricks_catalog" "sandbox" {
+			name           = "sandbox{var.RANDOM}"
+			comment        = "this catalog is managed by terraform"
+			properties     = {
+				purpose = "testing"
+			}
+		}`,
+	}, step{
+		Template: `
+		resource "databricks_catalog" "sandbox" {
+			name           = "sandbox{var.RANDOM}"
+			isolation_mode = "ISOLATED"
+			comment        = "this catalog is managed by terraform"
+			properties     = {
+				purpose = "testing"
+			}
+		}`,
+	}, step{
+		Template: `
+		resource "databricks_catalog" "sandbox" {
+			name           = "sandbox{var.RANDOM}"
+			isolation_mode = "OPEN"
+			comment        = "this catalog is managed by terraform"
+			properties     = {
+				purpose = "testing"
+			}
+		}`,
+	})
+}

--- a/internal/acceptance/catalog_test.go
+++ b/internal/acceptance/catalog_test.go
@@ -21,7 +21,7 @@ func TestUcAccCatalogIsolated(t *testing.T) {
 	unityWorkspaceLevel(t, step{
 		Template: `
 		resource "databricks_catalog" "sandbox" {
-			name           = "sandbox{var.RANDOM}"
+			name           = "sandbox{var.STICKY_RANDOM}"
 			comment        = "this catalog is managed by terraform"
 			properties     = {
 				purpose = "testing"
@@ -30,7 +30,7 @@ func TestUcAccCatalogIsolated(t *testing.T) {
 	}, step{
 		Template: `
 		resource "databricks_catalog" "sandbox" {
-			name           = "sandbox{var.RANDOM}"
+			name           = "sandbox{var.STICKY_RANDOM}"
 			isolation_mode = "ISOLATED"
 			comment        = "this catalog is managed by terraform"
 			properties     = {
@@ -40,7 +40,7 @@ func TestUcAccCatalogIsolated(t *testing.T) {
 	}, step{
 		Template: `
 		resource "databricks_catalog" "sandbox" {
-			name           = "sandbox{var.RANDOM}"
+			name           = "sandbox{var.STICKY_RANDOM}"
 			isolation_mode = "OPEN"
 			comment        = "this catalog is managed by terraform"
 			properties     = {

